File: api/internals/routes/stagiaire.go
package routes

import (
	"project-stage-pfa/internals/handlers"

	"github.com/gin-gonic/gin"
)

func RegisterStagiaireRoutes(r *gin.Engine) {

	stagiaire := r.Group("/stagiaire")

	// Basic routes
	stagiaire.GET("/", handlers.GetAllStagiaires())
	stagiaire.POST("/", handlers.CreateStagiaire())
	stagiaire.GET("/:id", handlers.GetStagiaireById())
	// presence total
	stagiaire.GET("/:id/presence/:id_stage", handlers.GetStagiaireAllPresence())
	stagiaire.POST("/:id/presence/:id_stage/today", handlers.MarquerStagiairePresentAujourdui())

	// demande-stage
	stagiaire.POST("/:id/demande-stage", handlers.CreateDemandeForStagiaire())

	// Update and Delete routes
	stagiaire.PUT("/:id", handlers.UpdateStagiaire())
	stagiaire.DELETE("/:id", handlers.DeleteStagiaire())

	// Nested group for routes that use a stagiaire ID
}

File: frontend/src/lib/api/stagiaire.js
// API base URL - adjust this to match your Go server
const API_BASE_URL = "http://localhost:8080"; // Change port if needed

// Generic API fetch function
const apiRequest = async (endpoint, options = {}) => {
  const url = `${API_BASE_URL}${endpoint}`;

  const config = {
    headers: {
      "Content-Type": "application/json",
      ...options.headers,
    },
    ...options,
  };

  try {
    const response = await fetch(url, config);

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(
        errorData.message || `HTTP error! status: ${response.status}`
      );
    }

    return await response.json();
  } catch (error) {
    console.error(`API request failed for ${endpoint}:`, error);
    throw error;
  }
};

// Stagiaire API functions
const stagiaireApi = {
  // Get all stagiaires
  createStagiaire: (stagiaireData) =>
    apiRequest("/stagiaire", {
      method: "POST",
      body: JSON.stringify(stagiaireData),
    }),
  getAllStagiaires: () => apiRequest("/stagiaire/"),

  // Get stagiaire by ID
  getStagiaireById: (id) => apiRequest(`/stagiaire/${id}`),

  // Get stagiaire presence for a specific stage
  getStagiairePresence: (stagiaireId, stageId) => 
    apiRequest(`/stagiaire/${stagiaireId}/presence/${stageId}`),

  // Mark stagiaire present for today
  markStagiairePresent: (stagiaireId, stageId) => 
    apiRequest(`/stagiaire/${stagiaireId}/presence/${stageId}/today`, {
      method: "POST",
    }),

  // Create demande stage for stagiaire
  createDemandeForStagiaire: (stagiaireId, demandeData) => 
    apiRequest(`/stagiaire/${stagiaireId}/demande-stage`, {
      method: "POST",
      body: JSON.stringify(demandeData),
    }),

  // Update stagiaire
  updateStagiaire: (id, stagiaireData) => apiRequest(`/stagiaire/${id}`, {
    method: "PUT",
    body: JSON.stringify(stagiaireData),
  }),

  // Delete stagiaire
  deleteStagiaire: (id) => apiRequest(`/stagiaire/${id}`, {
    method: "DELETE",
  }),
};

export default stagiaireApi;
